packages/js/components/src/date-time-picker-control/date-time-picker-control.tsx

```diff
@@ -174,72 +174,118 @@ export const DateTimePickerControl: React.FC< DateTimePickerControlProps > = ( {
 	// - [useDebounce hook loses function calls if the dependency changes](https://github.com/WordPress/gutenberg/issues/35505)
 	// - [useMemoOne and useCallbackOne](https://github.com/alexreardon/use-memo-one)
 
-	const onChangePropFunctionRef = useRef<
+	const onChangeRef = useRef<
 		DateTimePickerControlOnChangeHandler | undefined
 	>();
 	useLayoutEffect( () => {
-		onChangePropFunctionRef.current = onChange;
+		onChangeRef.current = onChange;
 	}, [ onChange ] );
 
-	function inputStringChangeHandlerFunction(
-		newInputString: string,
-		fireOnChange: boolean
-	) {
-		if ( ! isMounted.current ) return;
+	function setInputStringWithMoment( dateTime: Moment ) {
+		const newInputString = dateTime.isValid()
+			? formatMoment( dateTime )
+			: dateTime.creationData().input?.toString() || '';
+
+		setInputString( newInputString );
+	}
+
+	function callOnChange( dateTime: Moment ) {
+		const valueToSend = dateTime.isValid()
+			? formatMomentIso( dateTime )
+			: dateTime.creationData().input?.toString() || '';
+
+		if ( typeof onChangeRef.current === 'function' ) {
+			onChangeRef.current( valueToSend, dateTime.isValid() );
+		}
+	}
 
-		let newDateTime = parseMoment( newInputString );
+	function updateState(
+		newDateTimeString: string,
+		isLocalTime: boolean,
+		isManuallyTypedInput: boolean,
+		maybeFireOnChange: boolean
+	) {
+		const previousLastValidDate = lastValidDate.current;
+		let newDateTime = isManuallyTypedInput
+			? parseMoment( newDateTimeString )
+			: parseMomentIso( newDateTimeString, isLocalTime );
 		const isValid = newDateTime.isValid();
 
 		if ( isValid ) {
 			newDateTime = maybeForceTime( newDateTime );
-			setLastValidDate( newDateTime );
+			lastValidDate.current = newDateTime;
+		}
+
+		if ( isManuallyTypedInput ) {
+			// We don't want to reformat what the user typed in
+			setInputString( newDateTimeString );
+		} else {
+			setInputStringWithMoment( newDateTime );
 		}
 
 		if (
-			fireOnChange &&
-			typeof onChangePropFunctionRef.current === 'function'
+			maybeFireOnChange &&
```

Comment:
Not sure if this entirely possible given some of the logic, but could this be moved into a `useEffect`?

```
useEffect( () => {
  if ( isInitialUpdate.current ) {
    return;
  }
  if ( newDateTime.isSame( previousLastValidDate ) ) {
    return;
  }

  callOnChange( newDateTime );
  isInitialUpdate.current = false;
}, [ newDateTime ] );

Reply:
I tried a `useEffect` approach where I called `onChange` from within it, but it had some really nasty complications with both `currentDate` and `inputString` potentially changing. I fought with it for quite a while and then arrived at a much better solution where I'm not using `useEffect` at all to handling the `currentDate`, `inputString`, and `onChange`. I think it is now much more maintainable and readable.